### DISCUSSION TITLE
Orchestrator function backup record in Azure Table stays in Running state

### DISCUSSION THREAD

**Discussion by romangobrey** [Azure/azure-functions-durable-extension#1797] [Q&A] [answered]
Hello @cgillum,

I have a simple trigger function and orchestrator. 
Noticed that if an orchestrator function has async in its signature along with Task as an output or generic Task and also has something in the body with await the backup record in azure table stays in Running state... Is it normal? I guess not. As soon as the function ends (exits normally) execution instance is not marked as Completed. If I change the function signature to be sync (e.g. returns void and there is not async word) and function ends - this instance is marked as Completed immediately. The same with async and Task in signature but without code that uses await in the body - also works as expected. 

```
public static class Class1
{
    [FunctionName(nameof(TriggerAsync))]
    public static async Task TriggerAsync(
        [DurableClient] IDurableOrchestrationClient orchestrationClient,
        [DurableClient] IDurableEntityClient entityClient,
        [TimerTrigger("0 */1 * * * *")] TimerInfo timer,
        ILogger logger)
    {
        logger.LogWarning("Trigger. Starting new orchestration");
        await orchestrationClient.StartNewAsync(nameof(OrchestrationAsync));
        logger.LogWarning("Trigger. Started new orchestration");
    }

    [FunctionName(nameof(OrchestrationAsync))]
    public static async Task OrchestrationAsync(
        [OrchestrationTrigger] IDurableOrchestrationContext context,
        [DurableClient] IDurableOrchestrationClient starter,
        ILogger logger)
    {
        try
        {
            logger.LogWarning($"Orchestration. Doing something...");
            await starter.GetStatusAsync(context.InstanceId);
        }
        catch
        {
            logger.LogWarning($"Orchestration. Error");
            throw;
        }

        logger.LogWarning($"Orchestration. Done");
    }
}
```

Thank you for help!

**Comment by ConnorMcMahon** [comment, marked as answer]
Your orchestration function is violating [our code restraints](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-code-constraints#using-deterministic-apis), specifically the "async APIs" code restraint. The only async calls you should make inside your orchestration are on `Task`s created via `IDurableOrchestrationContext` APIs (or helper methods that themselves follow the orchestration code restraints).

If you want to make asynchronous calls, especially those doing IO like `IDurableOrchestrationClient`, you need to wrap those calls inside of an activity function.

Our Durable Functions analyzer should have generated warnings for this. Do you see these warnings, and if not, what version of the nuget package are you using?

**Reply by romangobrey** [reply]
@ConnorMcMahon ,
Thanks for quick reply and clarification.

Strange thing, but I tried to inject IDurableOrchestrationClient into activity func and no luck, dependency was not resolved. Gave it another try and it worked... Magic :)

Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers is added to the project however I don't see any suggestions to adjust the code, should I configure it somehow in Visual Studio.

Thanks again!

**Reply by davidmrdavid** [reply]
@amdeel might be able to correct me here, but I believe all you need to do to enable the Analyzer is to enable full solution analysis in your Visual Studio settings. This can be done in Tools -> Options -> Text Editor -> C# -> Advanced --> "Entire solution".

**Reply by romangobrey** [reply]
@davidmrdavid thanks for pointing, enabled the setting however I still don't see any hints? from analyzer.

Attached is an example for the issue described above from finally block in orchestrator
The question is: where should I see/find these hints from analyzer? 

![Screenshot 2021-04-26 114345](https://user-images.githubusercontent.com/19572565/116054931-e0a90800-a684-11eb-9907-afe4a7b29333.png)

Thanks again!

**Reply by amdeel** [reply]
David was correct about how to enable a lot of the analyzers. The analyzers will flag errors as it sees them. You can make sure the full solution setting is enabled correctly by attempting to input an incorrect value into your CallActivityAsync method, like an incorrect input type, for example.

At one point I added an analyzer that checked for any awaits in an orchestrator and flagged them if they weren't used with a Durable type, but that analyzer got scrapped because I ran into too many problems with it. Our analyzers currently check for usage of specific types instead.

So here, you've run into a situation where we aren't checking for your specific type. This would be something to improve about our analyzer. I recommend brushing up on our code constraints here: https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-code-constraints.

Remember: Async APIs are not allowed.

**Reply by romangobrey** [reply]
@amdeel Thank you!
